### ISSUE TITLE
Co-process hooks

### ISSUE THREAD

**Issue by tgross** [joyent/containerpilot#175]
A frequent request we've received is "how do I run multiple processes with ContainerPilot?" (ex https://github.com/joyent/containerpilot/issues/157 but in other places too). The primary reason we've balked at having ContainerPilot act as a supervisor is because it makes the semantics of restarts and signal handling difficult with relationship to some higher-level scheduler. In other words, if one of the applications stops do we stop the whole container or restart the application?

The topology problems with Consul deployments on Triton and PaaS-like environment described in https://github.com/joyent/containerpilot/issues/162 demonstrates a use case for co-processes where the main application is monitored by ContainerPilot but we want a persistent (rather than periodic) process to be run alongside it. This would also match nicely with running `consul-template` persistently as in #157. Additionally, in these scenarios it's clear that the extra process is intended to be a co-process -- it's not the main application. This means we don't have to worry about passing along signals, etc. other than reaping children.

The proposal here is this:
- allow for a `coprocess` field in the configuration which takes an array of commands
- start each coprocess immediately after the `preStart` but before starting our main application
- SIGTERM ourselves if any of the co-processes dies

cc @misterbisson @justenwalker @dekobon

**Comment by justenwalker** [comment]
Need to discuss more about exit codes, stderr/out. Also if it expected by
behavior that when the coprocess dies, the container dies - I think most
people want it to be restarted.

**Comment by tgross** [comment]
> Need to discuss more about exit codes, stderr/out.
> Also if it expected by behavior that when the coprocess dies, the container dies - I think most people want it to be restarted.

All our lifecycle hooks feed stdout/stderr thru the logger and we lose their exit codes.  As far as restarts, the co-process is secondary but what if the co-process can't be started because it's been misconfigured? We'd just end up restarting continuously and the main process might be in a bad state as a result.

**Comment by justenwalker** [comment]
I wonder how the log will look with coprocessor stdout interleaved with the main app

**Comment by tgross** [comment]
Funky, I imagine... the chance that the coprocess and the main app have similar logging formats is unfortunately pretty low. This might make it a pain for log parsing downstream.

**Comment by tgross** [comment]
> but what if the co-process can't be started because it's been misconfigured? We'd just end up restarting continuously and the main process might be in a bad state as a result.

A possible option to solve this dilemma could be to have a configurable number of restarts on the coprocess:

```json
{"coprocess":
  [
    {
      "cmd": "consul-template",
      "restarts": 5
    }
  ]
}
```

And then once those restarts are used up, we stop the whole container?

**Comment by tgross** [comment]
Note that I'm intentionally leaving out some extra complexity there like backoff timing on restarts because I really don't want to include that level of logic in ContainerPilot -- it's too application specific. If someone has that kind of problem they should use a real supervisor.

**Comment by tgross** [comment]
@justenwalker seeing as how there's no objection in principle here and we're now just going over implementation details, I'm going to put together a quick prototype PR. It'll be heavily borrowing from the work you did on `tasks`.

**Comment by tgross** [comment]
Draft implementation for discussion is here: https://github.com/joyent/containerpilot/pull/176

**Comment by tgross** [comment]
> seeing as how there's no objection in principle here

Turns out there's some internal objections on this. Lemme work that out in terms of direction before we invest any more effort on that.

**Comment by justenwalker** [comment]
> @justenwalker seeing as how there's no objection in principle here and we're now just going over implementation details, I'm going to put together a quick prototype PR. It'll be heavily borrowing from the work you did on tasks.

Before we get too far into this.  Is there a specific thing that adding this feature gives us that simply using `runit` or another supervisor does not?

Ones I think are valid:
- Exit code of container == exit code of main application
- Signals forwarded correctly to main application
- Log redirection to stdout/stderr correctly

although perhaps `runit` or other supervisors take care of that to some degree?

**Comment by tgross** [comment]
> Before we get too far into this. Is there a specific thing that adding this feature gives us that simply using runit or another supervisor does not?

The items you've listed are all the reasons why, and also happen to be the major objections to trying to run multiprocess Docker containers in general -- these behaviors all make it hard for any upstack scheduler to get the correct behavior.

[My fork of autopilotpattern/nginx](https://github.com/tgross/nginx-autopilotpattern/tree/multiprocess) illustrates the problems: you need a [`postStop`](https://github.com/tgross/nginx-autopilotpattern/blob/f7b34f8d4e7ae6fceefb028d0e480af00a6e880c/bin/manage.sh#L16-L19) behavior to kill the container, and we totally lose the log redirection of stdout/stderr from the main application and SIGHUP behavior for ContainerPilot. No supervisor I know of handles these problems, which is partly why we built ContainerPilot in the first place.

**Comment by tgross** [comment]
Some internal discussion on this proposal, with some minor editing for clarity.

@dekobon said:
> I like your proposal. In fact, it has a lot of uses. Setting up runnit or any multi-process manager in Docker is a pain. As for logs, I think we can come up with some solutions. For one, just being able to disable log output for a coprocess will be useful.

@tgross said:
> There's no reason to treat them any differently than the 7 other coprocesses we run in terms of logs.
> It's not a free lunch though. If the co-process needs to be configured in a way that can't be done at start time (obvious example `consul join`) then the end-user is going to have to add stuff into `health` or `tasks` to do a first-run configure. none of that complexity goes away.

@dekobon said:
> I wasn't expecting it to. However, if I needed the whole enchilada I would use a multi-process manager, but just having the freedom to run simple coprocesses in a well-defined startup sequence is amazing.

@tgross
> One concern I have is that we're going to end up with lots of this: https://github.com/joyent/containerpilot/issues/173

@dekobon said:
> That is a rough one. In fact things like tomcat will behave rather pathologically when you are trying to determine that.

@tgross
> Yuuup. Applications just take a certain non-zero amount of time to start. Which you get even with a "real" supervisor if there are dependencies

@dekobon said:
> Traditionally, it was the load balancer's job to make that determination (if a web app)

@tgross
> Well sure, you just don't mark yourself healthy until you're ready to work

@dekobon said:
> Some of the crazy process manager scripts that I've seen will keep pinging on a local socket to determine a state change. However, that's all out of scope. Come to think of it - for this particular problem it should be up to the implementor to write a proper health check that will tell them the answer.

**Comment by tgross** [comment]
Although the implementation is still WIP, here's a demonstration of its use by running a Consul agent within the container as a coprocess, for Nginx: https://github.com/tgross/nginx-autopilotpattern/tree/coprocess.

In particular the diff for the [containerpilot.json](https://github.com/tgross/nginx-autopilotpattern/commit/2f6678e1386f808190c75eb347800415be4fd502) file shows the only real changes you need to the application to make something like https://github.com/joyent/containerpilot/issues/162 work out:
- point the Consul configuration for ContainerPilot to localhost
- add a `coprocesses` stanza to fire up Consul, giving it a name for the `-join` argument that we can get via env var interpolation

I've smoke-tested this both on Docker for Mac and Triton.

**Comment by tgross** [comment]
Released in https://github.com/joyent/containerpilot/releases/tag/2.3.0
